File: DSA/SakshamKashyap_590014183/day52/day52_Question_3.py
def trap_rain_water(height):
    if not height:
        return 0

    left, right = 0, len(height) - 1
    left_max, right_max = height[left], height[right]
    trapped_water = 0

    while left < right:
        if left_max < right_max:
            left += 1
            left_max = max(left_max, height[left])
            trapped_water += max(0, left_max - height[left])
        else:
            right -= 1
            right_max = max(right_max, height[right])
            trapped_water += max(0, right_max - height[right])

    return trapped_water

# Example usage:
height1 = [0,1,0,2,1,0,1,3,2,1,2,1]
print(trap_rain_water(height1))  # Output: 6

height2 = [4,2,0,3,2,5]
print(trap_rain_water(height2))  # Output: 9

File: DSA/Gautam_Shyamnani_590013078/Day52Question3.cpp
#include <iostream>
#include <vector>

int trapRainWater(const std::vector<int>& height) {
    int left = 0, right = height.size() - 1;
    int left_max = 0, right_max = 0;
    int trapped_water = 0;

    while (left < right) {
        if (height[left] < height[right]) {
            if (height[left] >= left_max)
                left_max = height[left];
            else
                trapped_water += left_max - height[left];
            ++left;
        } else {
            if (height[right] >= right_max)
                right_max = height[right];
            else
                trapped_water += right_max - height[right];
            --right;
        }
    }

    return trapped_water;
}

int main() {
    std::vector<int> height1 = {0,1,0,2,1,0,1,3,2,1,2,1};
    std::cout << "Trapped water: " << trapRainWater(height1) << std::endl; // Output: 6

    std::vector<int> height2 = {4,2,0,3,2,5};
    std::cout << "Trapped water: " << trapRainWater(height2) << std::endl; // Output: 9

    return 0;
}
